crypto/bio/bio_print.c

```diff
@@ -8,947 +8,70 @@
-                return 0;
-            if (*currlen > 0) {
-                if (!ossl_assert(*sbuffer != NULL))
-                    return 0;
-                memcpy(*buffer, *sbuffer, *currlen);
-            }
-            *sbuffer = NULL;
-        } else {
-            char *tmpbuf;
-
-            tmpbuf = OPENSSL_realloc(*buffer, *maxlen);
-            if (tmpbuf == NULL)
-                return 0;
-            *buffer = tmpbuf;
-        }
-    }
-
-    if (*currlen < *maxlen) {
-        if (*sbuffer)
-            (*sbuffer)[(*currlen)++] = (char)c;
-        else
-            (*buffer)[(*currlen)++] = (char)c;
-    }
-
-    return 1;
-}
-
-/***************************************************************************/
-
 int BIO_printf(BIO *bio, const char *format, ...)
 {
     va_list args;
     int ret;
 
     va_start(args, format);
-
     ret = BIO_vprintf(bio, format, args);
-
     va_end(args);
     return ret;
 }
 
 int BIO_vprintf(BIO *bio, const char *format, va_list args)
 {
-    int ret;
-    size_t retlen;
-    char hugebuf[1024 * 2];     /* Was previously 10k, which is unreasonable
-                                 * in small-stack environments, like threads
-                                 * or DOS programs. */
-    char *hugebufp = hugebuf;
-    size_t hugebufsize = sizeof(hugebuf);
-    char *dynbuf = NULL;
-    int ignored;
-
-    dynbuf = NULL;
-    if (!_dopr(&hugebufp, &dynbuf, &hugebufsize, &retlen, &ignored, format,
-                args)) {
-        OPENSSL_free(dynbuf);
-        return -1;
-    }
-    if (dynbuf) {
-        ret = BIO_write(bio, dynbuf, (int)retlen);
-        OPENSSL_free(dynbuf);
-    } else {
-        ret = BIO_write(bio, hugebuf, (int)retlen);
-    }
-    return ret;
+    size_t ret;
+    int i = -1;
+    char buff[1024];
+    va_list copy;
+
+    /* Does it fit in the fixed buffer? */
+    va_copy(copy, args);
+    ret = vsnprintf(buff, sizeof(buff), format, args);
+    if (ret < sizeof(buff))
+	i = BIO_write(bio, buff, ret);
+    else {
+	/* Allocate enough space and try again. */
+	size_t size = ret + 1;
+	char *p = OPENSSL_malloc(size);
+
+	ret = vsnprintf(p, size, format, copy);
+	if (ret < size)
+	    i = BIO_write(bio, p, ret);
+	OPENSSL_free(p);
+    }
+    va_end(copy);
+    return i;
 }
 
 /*
- * As snprintf is not available everywhere, we provide our own
- * implementation. This function has nothing to do with BIOs, but it's
- * closely related to BIO_printf, and we need *some* name prefix ... (XXX the
- * function should be renamed, but to what?)
+ * This is a wrapper around snprintf, but returns -1 if the
+ * buffer is too small (documented behavior).
  */
-int BIO_snprintf(char *buf, size_t n, const char *format, ...)
+int BIO_snprintf(char *buff, size_t n, const char *format, ...)
 {
     va_list args;
-    int ret;
+    size_t ret;
 
     va_start(args, format);
-
-    ret = BIO_vsnprintf(buf, n, format, args);
-
+    ret = vsnprintf(buff, n, format, args);
```

Comment:
vsnprintf doesn't return a size_t.

Reply:
aargh.  how did I miss that?  thanks, fixup commit pushed (already built cleanly on my machine)

Reply:
BIO_snprintf() and BIO_vsnprintf() still have size_t

I'm really not a fan of these wrappers. As we have seen in the last pull request they also [hide bugs](https://github.com/quictls/quictls/pull/132#discussion_r1871244988) since it's annoying to add the relevant compiler attributes portably.

Reply:
I am not a fan of them either. A major problem is that the BIO functions have a different return value, -1 on error rather than the needed size. Eventually we should clean up our code.